src/Data.Json.JTable.Internal.purs

```diff
@@ -0,0 +1,214 @@
+widthOfPrimTuple :: JPath -> [Json] -> Maybe Number
+widthOfPrimTuple path ja = 
+  path !! 0 *> ja !! 1 *> let
+    types = ja <#> foldJson (\_->0) (\_->1) (\_->2) (\_->3) (\_->4) (\_->4)
+    not_same = length (nub types) /= 1
+    all_prim = all ((/=) 4) types
+    in (all_prim && (not_same || length ja == 2)) `so` length ja
+
+
+-- add child to tree, unify if exists
+tMergeArray :: Tree -> Tree -> Tree
+tMergeArray (T p w h k) nt@(T np nw nh nk) =
+  let i = findIndex (\n -> last np == last (n # tPath)) k in case k !! i of
+    Just t2@(T p2 w2 h2 k2) -> case foldl tMergeArray t2 nk of 
+      (T _ w2' h2' k2') -> let k' = updateAt i (T p2 w2' h2' k2') k
+                               w' = w - w2 + w2'
+                               h' = max h (h2' + 1)
+                           in T p w' h' k'
+    Nothing -> let w' = if null k then nw else w+nw
+                   h' = max h $ nh + 1
+                   k' = snoc k nt
+               in T p w' h' k'
+
+-- produce a tree of header data from json
+tFromJson :: JPath -> Json -> Tree
+tFromJson path json = let
+  obj = toObject json <#> \jo ->
+    if M.isEmpty jo then T path 1 0 []    
+    else let k = M.toList jo <#> uncurry \l j -> tFromJson (snoc path l) j
+             w = k <#> tWidth # foldl (+) 0
+             h = 1 + foldl max 0 (k <#> tHeight)
+         in T path w h k
+  m = (>>=) (toArray json)
+  tuple = m \a -> (widthOfPrimTuple path a) <#> \n -> T path n 0 []
+  array = m \a -> let
+    ts = a <#> tFromJson path
+    t' = foldl tMergeArray (T path 0 0 []) (ts >>= tKids)
+    w' = t' # tWidth
+    mw = ((head $ nub (ts <#> tWidth)) <#> max w') <|> (Just $ max 1 w')
+    in mw <#> \w -> T path w (t' # tHeight) (t' # tKids) 
+  prim = T path 1 0 []
+  in prim `fromMaybe` (obj <|> tuple <|> array)
+
+
+-- merge table segments for each key of an object into one
+cMergeObj :: [(Tuple Number Table)] -> Table
+cMergeObj rss = let
+  maxh = rss <#> snd <#> length # foldl max 0
+  in (0 .. maxh-1) <#> \n -> rss >>= uncurry \w rs ->
+    if length rs == 1 then if n == 0 
+                           then AU.head rs <#> \(C c w h j) -> C c w maxh j
+                           else [] `fromMaybe` (rs !! n)
+    else [C (JCursorTop) w 1 primNull] `fromMaybe` (rs !! n)
+
+-- maybe merge a tuple of objects into a table segment
+mergeObjTuple ::Tree -> JCursor -> [Json] -> Maybe Table
+mergeObjTuple t@(T p w h k) c ja = head ja *> do
+  jos <- for ja toObject
+  let keyss = jos <#> M.keys
+  let all_keys = concat keyss
+  so ((length all_keys) == (length $ nub all_keys)) $ 
+    cMergeObj $ k <#> \(t'@(T p' w' _ _)) -> let 
+      label = AU.last p'
+      i = findIndex (elem label) keyss
+      j = jnull `fromMaybe` (jos !! i >>= M.lookup label)
+      in Tuple w' (cFromJson t' (downField label $ downIndex i c) j)
+
+-- produce data table from json, according to header tree
+cFromJson :: Tree -> JCursor -> Json -> Table
+cFromJson t@(T p w h k) c json = let
```

Comment:
Again, I always try to avoid the `Maybe` monad when possible, because code so written is harder to reason about because any `Nothing` anywhere in a chain can cause the whole thing to be `Nothing`. It's sort of the Javascript of functional programming. In the case of this particular function, we can eliminate a lot of this by using `foldJson`, and combining the width and tuple computation. Also making `toPrim` type-accurate will have some effects on this function (cleaning it up, I believe, and making it easier to reason about).